compiler/rustc_query_system/src/query/plumbing.rs

```diff
@@ -300,7 +300,18 @@ where
     match result {
         Ok(()) => {
             let Some((v, index)) = query.query_cache(qcx).lookup(&key) else {
-                cold_path(|| panic!("value must be in cache after waiting"))
+                cold_path(|| {
+                    // We didn't find the query result in the query cache. Check if it was
+                    // poisoned due to a panic instead.
+                    let lock = query.query_state(qcx).active.get_shard_by_value(&key).lock();
+                    match lock.get(&key) {
+                        // The query we waited on panicked. Continue unwinding here.
+                        Some(QueryResult::Poisoned) => FatalError.raise(),
+                        _ => panic!(
+                            "query result must in the cache or the query must be poisoned after a wait"
+                        ),
+                    }
```

Comment:
The only possible course of action is unwinding. Should we replace the match by an assertion?

```suggestion
                    // We didn't find the query result in the query cache. Check if it was
                    // poisoned due to a panic instead.
                    debug_assert!(
                        matches!(
                            query.query_state(qcx).active.get_shard_by_value(&key).lock().get(&key),
                            Some(QueryResult::Poisoned),
                        ),
                        "query result must in the cache or the query must be poisoned after a wait",
                    );
                    // The query we waited on panicked. Continue unwinding here.
                    FatalError.raise()
```

Reply:
I feel like the current version makes it more clear that we should only call `FatalError.raise()` once we observe the `QueryResult::Poisoned` case. I don't think I want to downgrade to a `debug_assert` without some more testing of the parallel compiler. This path is also not terribly hot.